KC-1307: Add support for Raft in KafkaClusterTestKit

aloknnikhil opened the pull request:
- Adds new tests that create and list topic(s) with more than 1 partition
- Refactor RaftConfig to use a Controller Quorum Voters Future to resolve the active connect string

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cmccabe commented:
Can you rebase this on top of kip-500, so that there are no merges in the branch?  It's tough to figure out what is going on with the merges.

cmccabe commented:
Preliminary thoughts (I can say more once you rebase): we need to get away from using so many `CompletableFuture` objects in initialization.  It's really kind of a hack.  MetaLogManager should definitely not be a CompletableFuture.

The log manager doesn't do anything until initialize() is called so we should probably just pass the quorum port config in as part of the initialize() function.  That will simplify this whole thing, I think.

aloknnikhil commented:
> Can you rebase this on top of kip-500, so that there are no merges in the branch? It's tough to figure out what is going on with the merges.

Ack. Done. Cleaned up and squashed a few commits too

aloknnikhil commented:
> Preliminary thoughts (I can say more once you rebase): we need to get away from using so many `CompletableFuture` objects in initialization. It's really kind of a hack. MetaLogManager should definitely not be a CompletableFuture.
> 
> The log manager doesn't do anything until initialize() is called so we should probably just pass the quorum port config in as part of the initialize() function. That will simplify this whole thing, I think.

Yea, I agree with this. I was not happy with the number of futures being passed around either. I think what you're suggesting makes sense and should clean it up. Let me take a stab at it now

cmccabe commented:
Thanks, @aloknnikhil .  I still think this change is a lot bigger than it needs to be.

I think there's basically two paths here:
1. pass the `RaftManager` constructor a `CompletableFuture[String]` for quorum voters.  Then basically nothing needs to change except when you create `RaftManager#raftConfig`, you set `controller.quorum.voters` to this.

2. Change `RaftManager#startup` so that it takes a string containing the voter information.  Then you don't need a future in the constructor.  This disadvantage is now the voter nodes are mutable state.  This is annoying, but probably not too bad.

If you do #1 then that's practically all you have to do.  Maybe it's better to start with that and go from there.

aloknnikhil commented:
> Thanks, @aloknnikhil . I still think this change is a lot bigger than it needs to be.
> 
> I think there's basically two paths here:
> 
> 1. pass the `RaftManager` constructor a `CompletableFuture[String]` for quorum voters.  Then basically nothing needs to change except when you create `RaftManager#raftConfig`, you set `controller.quorum.voters` to this.
> 2. Change `RaftManager#startup` so that it takes a string containing the voter information.  Then you don't need a future in the constructor.  This disadvantage is now the voter nodes are mutable state.  This is annoying, but probably not too bad.
> 
> If you do #1 then that's practically all you have to do. Maybe it's better to start with that and go from there.

#1 is a bit more nuanced than that actually. RaftManager exports the MetaLogManager (shim today) that the Kip500Controller uses as the MetaLogManager for the QuorumController. If we delay the creation of the MetaLogManager until after startup (which we have to considering it depends on the RaftClient which in turn depends on the RaftConfig and the voters string future), then we end up with a cyclic dependency without the changes in the PR. I am not sure just passing in a future string would work then. Does that make sense?

mumrah commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -16,15 +16,10 @@
 import kafka.raft.KafkaRaftManager.RaftIoThread
 import kafka.server.KafkaServer.ControllerRole
-import kafka.server.{BrokerTopicStats, KafkaBroker, KafkaConfig, LogDirFailureChannel, MetaProperties}
+import kafka.server._
```
Do we normally allow wildcard imports? I think we might not

aloknnikhil replied:
Not sure. Checkstyle didn't complain. So, I assumed it was allowed. Also, it was IntelliJ's "Optimize Imports" that did this. I can go bump up the number of imports before replacing it with a wildcard if it's preferred to not have wildcards.

mumrah commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -200,26 +199,18 @@ public KafkaClusterTestKit build() throws Exception {
+                            OptionConverters.toScala(Optional.of(node.id())));
+                    TopicPartition metadataPartition = new TopicPartition(KafkaServer.metadataTopicName(), 0);
+                    KafkaRaftManager raftManager = new KafkaRaftManager(metaProperties, metadataPartition, config,
+                            Time.SYSTEM, new Metrics(), connectFutureManager.future);
```
nit: use the existing `time` that's already in scope here

aloknnikhil replied:
Good catch. Thanks!

mumrah commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -141,30 +146,23 @@ public KafkaClusterTestKit build() throws Exception {
+                            OptionConverters.toScala(Optional.of(node.id())));
+                    TopicPartition metadataPartition = new TopicPartition(KafkaServer.metadataTopicName(), 0);
+                    KafkaRaftManager raftManager = new KafkaRaftManager(metaProperties, metadataPartition, config,
+                            Time.SYSTEM, new Metrics(), connectFutureManager.future);
```
nit: use `time`

aloknnikhil replied:
Good catch. Thanks!

mumrah commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -141,30 +146,23 @@ public KafkaClusterTestKit build() throws Exception {
+                            OptionConverters.toScala(Optional.of(node.id())));
+                    TopicPartition metadataPartition = new TopicPartition(KafkaServer.metadataTopicName(), 0);
+                    KafkaRaftManager raftManager = new KafkaRaftManager(metaProperties, metadataPartition, config,
+                            Time.SYSTEM, new Metrics(), connectFutureManager.future);
```
Do we want to share a Metrics instance between Kip500Controller and KafkaRaftManager?

aloknnikhil replied:
Good catch. Yea, it makes sense to use the same metrics instance here.

aloknnikhil commented on raft/src/main/java/org/apache/kafka/raft/RaftConfig.java:
```diff
@@ -61,14 +61,13 @@
+    public static final String QUORUM_RETRY_BACKOFF_MS_CONFIG = QUORUM_PREFIX +
         CommonClientConfigs.RETRY_BACKOFF_MS_CONFIG;
 
+    // TODO: Move these to KafkaConfig.scala and avoid duplication
```
Adding a TODO to track config migration to `KafkaConfig.scala` in another PR

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -1902,7 +1902,11 @@ public Long scheduleAppend(int epoch, List<T> records) {
     public LeaderAndEpoch leaderAndEpoch() {
-        return quorum.leaderAndEpoch();
+        LeaderAndEpoch leaderAndEpoch = new LeaderAndEpoch(OptionalInt.empty(), -1);
+        if (quorum != null) {
```
Hmm... This seems to suggest we are using `leaderAndEpoch` before initialization has completed. I am thinking how we can avoid that. If you look in `KafkaRaftClient.initialize()`, the only reason we need the addresses is to initialize the endpoints:
```java
        Map<Integer, InetSocketAddress> voterAddresses = raftConfig.quorumVoterConnections();
        for (Map.Entry<Integer, InetSocketAddress> voterAddressEntry : voterAddresses.entrySet()) {
            channel.updateEndpoint(voterAddressEntry.getKey(), voterAddressEntry.getValue());
        }
```
In fact, there's no reason why this must be done in `initialize`. It could also be done in `RaftManager.startup`. In fact, if you look at the handling inside `KafkaNetworkChannel`, we already handle the case when the destination address is not defined. So what we could do is something like this:

1. Initialize `controller.quorum.voters` using non-routable addresses. For example, we could use `1@0.0.0.0:0,2@0.0.0.0:0`. 
2. Add a method to `RaftConfig` which parses only the voter ids. This will be used to initialize the quorum state and such in `KafkaRaftClient.initialize`.
3. Move endpoint initialization out of `KafkaRaftClient.initialize` and into `RaftManager`. Basically when the future completes, we can call `KafkaNetworkChannel.updateEndpoint`.

aloknnikhil replied:
I think that makes sense. We can generate a quorum voters string using the number of controllers (which should be fixed and known up-front even in the TestKit) and the non-routable address

aloknnikhil replied:
We don't need to change anything in `RaftConfig.quorumVoterIds` actually if we pass in a connection string that is of the form you specified. It'll parse the voter IDs out as is. So, I think the only change we need is to handle the endpoint update outside in the RaftManager and also update the endpoints in the RaftConfig to keep it consistent

hachikuji replied:
Even better. I wasn't sure if `InetSocketAddress` validated that the ip was routable.

aloknnikhil commented on core/src/main/scala/kafka/server/KafkaServerStartable.scala:
```diff
@@ -48,7 +48,7 @@ class KafkaServerStartable(
       case _: Throwable =>
         // KafkaServerStartable.startup() calls shutdown() in case of exceptions, so we invoke `exit` to set the status code
-        fatal("Exiting Kafka.")
+        fatal("Exiting Kafka." )
```
Ah. Yes. Thank you!

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -307,31 +328,12 @@ private void fireHandleResign(int epoch) {
     }
 
     @Override
-    public void initialize(RaftConfig raftConfig) throws IOException {
```
Hmm.. This is a little different than what I had in mind. I thought we would leave `initialize(RaftConfig)` as it was and move the block below to update the endpoints into `RaftManager`. From a high level, we are saying that it is the responsibility of `RaftManager` to initialize the endpoints and not `RaftClient`, which is a little more convenient for testing. Also, I had hoped to keep the call to `QuorumState.initialize` here.

aloknnikhil replied:
So, these changes aren't final yet and I was attempting to keep the RaftConfig and the RaftClient consistent with the endpoint updates (I.e. Avoid calls to update endpoints with the RaftConfig and with the RaftClient). From a testing perspective I see your point.

For the `QuorumState.initialize()`, I agree.

hachikuji replied:
I do think it's nice that we can make `RaftConfig` final now by the way. What if we did the following?

1. Keep `RaftConfig` in constructor
2. Move call to `QuorumState.initialize` to `KafkaRaftClient.initialize`
3. Move `updateEndpoint` calls up to `RaftManager.

aloknnikhil replied:
Yup. I like that too. RaftConfig being final again was definitely a goal considering we only need to have the endpoints updated in any given case.

hachikuji commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -121,17 +122,25 @@ public Builder setConfigProp(String key, String value) {
             Map<Integer, Kip500Controller> controllers = new HashMap<>();
             Map<Integer, Kip500Broker> kip500Brokers = new HashMap<>();
+            Map<Integer, KafkaRaftManager> raftManagers = new HashMap<>();
+            String dummyQuorumVotersString = nodes.controllerNodes().keySet().stream().
```
At a high level, I think we should make `0.0.0.0:0` a valid address configuration from the perspective of `KafkaRaftClient`. We are telling the client that initialization of the endpoints will come at a later time and through some external mechanism. I think this is a generally useful capability even outside of testing. For example, we may eventually want to be able to initialize an observer from a `bootstrap.servers` configuration. In this case, discovery of the voter endpoints will come after we have (say) sent a `DescribeQuorum` to the cluster. We may also want to implement some of the more fancy bootstrapping options that etcd has, such as using an external service to facilitate address bootstrapping.

Anyway, my point is we should treat this as less of a hack and more using a supported capability. To that end, I think we should document the behavior and add tests for it (we can do this separately of course).

aloknnikhil replied:
Yea, I think that's a fair point. Delaying the raft voter configuration has other applications as you said and I think we should do this in another PR that can also be then upstreamed

hachikuji commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -128,9 +128,15 @@ class KafkaRaftManager(
 
   def startup(): Unit = {
+    // Wait for the controller quorum voters string to be set
+    raftConfig.updateQuorumVoters(controllerQuorumVotersFuture.get())
```
Related to one of my other comments, but do we really need to do this? I think we should treat `RaftConfig` as immutable and only specifying the initial configuration. We can say that `NetworkChannel` is the source of truth for the current endpoint information of the voters.

aloknnikhil replied:
Yea, fair enough. I was going off of the expectation that the RaftConfig would stay consistent with the current cluster state. But this would mean that the RaftConfig is more of a source of truth what the cluster was configured with at startup, if that makes sense. Fixed.

aloknnikhil commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -111,15 +114,12 @@ class KafkaRaftManager(
     if (leaderAndEpoch.leaderId.isPresent) {
       val leaderId = leaderAndEpoch.leaderId.getAsInt
-      voterNodes.find(_.id == leaderId)
+      val leaderAddress = raftConfig.quorumVoterConnections().asScala(leaderId)
```
Good catch! Fixed.

aloknnikhil commented on core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:
```diff
@@ -121,17 +122,25 @@ public Builder setConfigProp(String key, String value) {
             Map<Integer, Kip500Broker> kip500Brokers = new HashMap<>();
+            Map<Integer, KafkaRaftManager> raftManagers = new HashMap<>();
+            String dummyQuorumVotersString = nodes.controllerNodes().keySet().stream().
+                map(controllerNode -> String.format("%d@0.0.0.0:1337", controllerNode)).
```
Fixed.
